File: java/349.intersection-of-two-arrays.java
/*
 * @lc app=leetcode id=349 lang=java
 *
 * [349] Intersection of Two Arrays
 */
class Solution {

    // Assume a large amount of repeat elements
    public int[] intersection(int[] nums1, int[] nums2) {
        if (nums1 == null || nums2 == null) return new int[] {};
        Set<Integer> s1 = new HashSet<>();
        List<Integer> l = new LinkedList<>();
        for (int i = 0; i < nums1.length; ++i) s1.add(nums1[i]);
        for (int i = 0; i < nums2.length; ++i) {
            if (s1.contains(nums2[i])) {
                l.add(nums2[i]);
                s1.remove(nums2[i]);
            }
        }
        int[] res = new int[l.size()];
        for (int i = 0; i < res.length; ++i) {
            res[i] = l.remove(0);
        }
        return res;
    }
}

File: go/349.intersection-of-two-arrays.go
/*
 * @lc app=leetcode id=349 lang=golang
 *
 * [349] Intersection of Two Arrays
 */

// @lc code=start

type void struct{}

var member void

func intersection(nums1 []int, nums2 []int) []int {
	res := make([]int, 0)
	set := make(map[int]void)
	for _, v := range nums2 {
		set[v] = member
	}
	for _, v := range nums1 {
		_, b := set[v]
		if b {
			delete(set, v)
			res = append(res, v)
		}
	}
	return res
}

// @lc code=end

